### DISCUSSION TITLE
Setting HighDpi Causes Pop-ups To Look Strange

### DISCUSSION THREAD

**Discussion by slane-ck** [cefsharp/CefSharp#4190] [Q&A] [answered]
I'm implementing CefSharp version 103.0.80.0 in my Winforms .net framework 4.8 app. I need to enable high dpi because otherwise there's a black border on the top and left of the browser control. I put Cef.EnableHighDPISupport() by my app startup, and I also added an app.manifest file with the following:

    <asmv3:application>
        <asmv3:windowsSettings xmlns="http://schemas.microsoft.com/SMI/2005/WindowsSettings">
            <dpiAware>true/PM</dpiAware>
        </asmv3:windowsSettings>
    </asmv3:application>

The black border disappears, but all of the popup forms in the app look strange, for example textboxes and buttons are getting cut off. Is there any way to make the high dpi only affect the browser control, not the whole form/app?

I've gone through https://github.com/cefsharp/CefSharp/wiki/General-Usage#high-dpi-displayssupport several times but haven't been able to fix this. 

I've run the MinimalExample but since it only has the web browser as a control and no other popups, I couldn't see if it works the same way.

Thank you.

**Comment by amaitland** [comment, marked as answer]
>  I put Cef.EnableHighDPISupport() by my app startup, and I also added an app.manifest file with the following:

As you are using .Net 4.8 the suggested application is to configure DPI awareness in the app.config 

> IMPORTANT If you are targeting .Net 4.7 or above Microsoft reccommends configuring DPI Awareness via the app.config rather than app.manifest. Make sure to read [High DPI support in Windows Forms](https://docs.microsoft.com/en-us/dotnet/desktop/winforms/high-dpi-support-in-windows-forms?view=netframeworkdesktop-4.8) for further details from Microsoft.

As per https://github.com/cefsharp/CefSharp/wiki/General-Usage#appmanifest

Remove the app.manifest entry and the call to Cef.EnableHighDPISupport() and use the app.config approach detailed at [High DPI support in Windows Forms](https://docs.microsoft.com/en-us/dotnet/desktop/winforms/high-dpi-support-in-windows-forms?view=netframeworkdesktop-4.8)

> I've gone through https://github.com/cefsharp/CefSharp/wiki/General-Usage#high-dpi-displayssupport several times but haven't been able to fix this.

There's more to read below the WPF section, make sure to read https://github.com/cefsharp/CefSharp/wiki/General-Usage#high-dpi-additional-info

**Reply by slane-ck** [reply]
What's the difference between the Cef.EnableHighDPISupport() call and setting it in app.config? It's not 2 ways of doing the same thing?

**Reply by amaitland** [reply]
> It's not 2 ways of doing the same thing?

Not anymore. WinForms now has it's own custom DPI awareness implementation, follow the instructions at https://docs.microsoft.com/en-us/dotnet/desktop/winforms/high-dpi-support-in-windows-forms?view=netframeworkdesktop-4.8

If you have problems double check you've removed the entries from the app.manifest and deleted the Cef.EnableHighDPISupport() call, see https://developercommunity.visualstudio.com/t/Issues-with-per-monitor-DPI-aware-WinFor/1463154

**Reply by slane-ck** [reply]
Thank you for your response.

I've removed the app.manifest file (since it wasn't there before) and the Cef.EnableHighDPISupport() call. I added the following to the bottom of my app.config file:

    <compatibility xmlns="urn:schemas-microsoft-com:compatibility.v1">
        <application>
            
            <supportedOS Id="{8e0f7a12-bfb3-4fe8-b9a5-48fd50a15a9a}" />
        </application>
        <System.Windows.Forms.ApplicationConfigurationSection>
            <add key="DpiAwareness" value="PerMonitorV2" />
        </System.Windows.Forms.ApplicationConfigurationSection>
    </compatibility>

I also added the Application.EnableVisualStyles() in my startup. 

However, my application seems completely DPI-unaware and it has the block border around the control again.

Could it have to do with that all of the forms in the project are set to the default autoscalemode, which is font? All of the designers look like this:
Me.AutoScaleDimensions = New System.Drawing.SizeF(6.0!, 13.0!)
Me.AutoScaleMode = System.Windows.Forms.AutoScaleMode.Font

Or is that unconnected?

I've read through the links that you posted but I didn't see anything that helped this situation. 

Thank you for your time!

**Reply by amaitland** [reply]
> I've removed the app.manifest file (since it wasn't there before) and the Cef.EnableHighDPISupport() call. I added the following to the bottom of my app.config file:

That's incorrect, you still need an app.manifest as per https://docs.microsoft.com/en-us/dotnet/desktop/winforms/high-dpi-support-in-windows-forms?view=netframeworkdesktop-4.8

**Reply by amaitland** [reply]
> Could it have to do with that all of the forms in the project are set to the default autoscalemode, which is font? 

That's likely part of the problem.

**Reply by amaitland** [reply]
Upon further review there's a mistake in https://docs.microsoft.com/en-us/dotnet/desktop/winforms/high-dpi-support-in-windows-forms?view=netframeworkdesktop-4.8 

The app.config shouldn't have a compatibility node. Actual example at https://docs.microsoft.com/en-us/dotnet/framework/configure-apps/file-schema/winforms/

Looks like it was reported already https://github.com/dotnet/docs-desktop/issues/1485

**Reply by slane-ck** [reply]
Thank you so much for your time. 

I added the app.manifest file and removed the compatibility node from the app.config file. In the app.manifest file, there's an additional section that sets the application to DPI-aware. Do I uncomment that part as well? It looks as follows (I kept the longPathAware node commented out because I'm not sure what it is):
```
<application xmlns="urn:schemas-microsoft-com:asm.v3">
    <windowsSettings>
      <dpiAware xmlns="http://schemas.microsoft.com/SMI/2005/WindowsSettings">true</dpiAware>
      <!--<longPathAware xmlns="http://schemas.microsoft.com/SMI/2016/WindowsSettings">true</longPathAware>-->
    </windowsSettings>
  </application>
```

Also, regarding the autoscalemode of the forms, the default for Winforms is autoscalemode=System.Windows.Forms.AutoScaleMode.Font, so if it needs to be Dpi then that should be in your documentation. In any case, even after setting the autoscalemode to dpi the popup forms still aren't scaling properly and they look the same as they did when autoscalemode is set to Font. 

Is there anything else I can try?

Thank you so much!

**Reply by slane-ck** [reply]
I'm rethinking this and I think that the Winforms app that I'm trying to integrate CefSharp into has too many forms and is too complex to get the high DPI to work. I'd rather just leave the black border. Is there any way for me to get rid of the black border without using the high dpi setting? I tried changing the location of the control but it didn't help.

Thank you!

**Reply by amaitland** [reply]
> then that should be in your documentation 

If you feel there is a need for better documentation then please direct your request to **Microsoft**. They should be the ones providing better overall guidance on making applications DPI aware. It's not a `CefSharp` specific feature.

> There's more to read below the WPF section, make sure to read https://github.com/cefsharp/CefSharp/wiki/General-Usage#high-dpi-additional-info

If you wish to keep your application `unaware` then that's an option, read the link I've already provided you.

**Reply by slane-ck** [reply]
Yes, I've switched to using the disable-gpu option and it seems to be working well. What's the difference between that and disable-gpu-compositing? 

Thank you so much!

**Reply by amaitland** [reply]
If you have an app.manifest with the compatibility entries then should only need to disable-gpu-compositing

Best to leave GPU acceleration enabled wherever possible.

**Reply by slane-ck** [reply]
Got it, thank you for your time.
